cluster/connection_pool.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"crypto/tls"
+	"fmt"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pkg/errors"
+)
+
+const capacity = 1024
+
+type connectionPool struct {
+	cache     *lru.Cache
+	tlsConfig *tls.Config
+}
+
+func newConnectionPool(tlsClientCfg *tls.Config) (*connectionPool, error) {
+	cache, err := lru.NewWithEvict(
+		capacity, func(_ interface{}, value interface{}) {
+			conn, ok := value.(*tlsConn)
+			if ok {
+				_ = conn.Close()
+			}
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new LRU")
+	}
+	return &connectionPool{
+		cache:     cache,
+		tlsConfig: tlsClientCfg,
+	}, nil
+}
+
+// borrowConnection returns a *tlsConn from the pool. The connection does not
+// need to be returned to the pool because each connection has its own locking.
+func (pool *connectionPool) borrowConnection(addr string, timeout time.Duration) (*tlsConn, error) {
+	if pool.cache == nil {
+		return nil, errors.New("connection pool closed")
+	}
+	key := fmt.Sprintf("%s/%d", addr, int64(timeout))
+	value, exists := pool.cache.Get(key)
+	if exists {
+		conn, ok := value.(*tlsConn)
+		if ok && conn.alive() {
+			return conn, nil
+		}
+	}
+	conn, err := dialTLSConn(addr, timeout, pool.tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+	pool.cache.Add(key, conn)
```

Comment:
@csmarchbanks I updated this to use an lru cache. I removed the mutex from this file because the cache provides locking. However,  I'm wondering if I still need it between the `pool.cache.Get` and the `pool.cache.Add`. Thoughts?

Reply:
I think yes as two concurrent operations could each dial a TLS connection and then add the connection for the same key?

Reply:
You might be able to use `PeekOrAdd` to avoid a mutex, but it would also have a bit of extra complexity. Basically:
1. Do the get
2. If not found, dial a TLS connection
3. Use PeekOrAdd
4. If found, use the "old" connection returned by PeekOrAdd, if added use the created connection.

Whichever way looks cleaner is fine by me.

Reply:
I tried both ways. In the end, the mutex ended looking cleaner. 

The `PeekOrAdd` would have been nice if I didn't always need to check whether the connection is alive when I get it back from the cache.